pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,188 @@
+package logger
+
+import (
+	"fmt"
+	"go.uber.org/zap/zapcore"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type ECSErrorLogger struct {
+	ErrorFilePath    string
+	ErrorLogMaxAge   int
+	ErrorFileMaxSize int
+
+	queue     []zapcore.Entry
+	ticker    time.Ticker
+	errorFile *os.File
+	size      int
+}
+
+// ECSErrorReporter func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func (l *ECSErrorLogger) ECSErrorReporter() {
+	defer l.ticker.Stop()
+	for {
+		select {
+		// when new error add in the channel, this case will be selected and write it into error log file
+		case e := <-errorLogChannel:
+			l.Write(e)
+		// this case will run every x mins (x is configurable) to delete the error which already expired, error log max age
+		// depends on STATUS_MESSAGE_TTL
+		case <-l.ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("could not open error log file when write new error, err: %v", err)
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorFileMaxSize {
+		log.Printf("error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file when write new error, err: %v", err)
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("could not write error into error log file, err: %v", err)
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+	// delete the old errors if old errors plus new error exceed the max size
+	if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+		l.rotate(newErrorLog)
+	}
+
+	l.queue = append(l.queue, newError)
+	n, err := l.errorFile.Write(newErrorLog)
+	if err != nil {
+		log.Printf("could not write new error into error log file, err: %v", err)
+	}
+	l.size += n
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func (l *ECSErrorLogger) rotate(newErrorLog []byte) {
```

Comment:
This function doesn't actually use the newErrorLog I think? Like you just use the length- so make the argument be the length instead.